godot-core/src/builtin/collections/packed_array.rs

```diff
@@ -433,10 +445,39 @@ macro_rules! impl_packed_array {
             }
         }
 
+        #[doc = concat!("Creates a `", stringify!($PackedArray), "` from the given Rust array.")]
+        impl<const N: usize> From<[$Element; N]> for $PackedArray {
+            fn from(arr: [$Element; N]) -> Self {
+                let mut ret = Self::new();
+                if N == 0 {
+                    return ret;
+                }
+                ret.resize(N);
+                // SAFETY: Data is moved and the original array is forgotten.
+                unsafe {
+                    let arr = std::mem::ManuallyDrop::new(arr);
+                    ret.move_from_slice(arr.as_ptr(), N);
+                }
+                ret
+            }
+        }
+
         #[doc = concat!("Creates a `", stringify!($PackedArray), "` from the given Rust vec.")]
         impl From<Vec<$Element>> for $PackedArray {
-            fn from(vec: Vec<$Element>) -> Self{
-                vec.into_iter().collect()
+            fn from(mut vec: Vec<$Element>) -> Self {
+                let mut ret = Self::new();
+                let len = vec.len();
+                if len == 0 {
+                    return ret;
+                }
+                ret.resize(len);
+                let ptr = vec.as_ptr();
+                // SAFETY: Data is moved and the original vector is forcibly set to empty.
```

Comment:
The `// SAFETY` assertion is not just a regular comment that explains what happens, but it needs to reason _why_ and _how_ the safety assumptions (specified in the function docs) are upheld.

Reply:
Would also be good to have an empty line before it.